couchdb/server/views.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+#
+import copy
+import logging
+
+from couchdb import json
+from couchdb.server.exceptions import QueryServerException, Error
+
+__all__ = ('map_doc',)
+
+log = logging.getLogger(__name__)
+
+
+def map_doc(server, doc):
+    """Applies available map functions to document.
+
+    :command: map_doc
+
+    :param server: Query server instance.
+    :type server: :class:`~couchdb.server.BaseQueryServer`
+
+    :param doc: Document object.
+    :type doc: dict
+
+    :return: List of key-value results for each applied map function.
+
+    :raises:
+        - :exc:`~couchdb.server.exceptions.Error`
+          If any Python exception occurs due mapping.
+    """
+    docid = doc.get('_id')
+    log.debug('Apply map functions to document `%s`:\n%s', docid, doc)
+    orig_doc = copy.deepcopy(doc)
+    map_results = []
+    _append = map_results.append
+    try:
+        for idx, func in enumerate(server.state['functions']):
+            # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
+            # Apply copy.deepcopy for `key` and `value` to fix this issue
+            _append([[key, value] for key, value in func(doc) or []])
```

Comment:
Patch V4
- prevent string being unpacked accidentally
- show the invalid value on err message
- more wordy(?) test cases
```diff
diff --git a/couchdb/server/views.py b/couchdb/server/views.py
index 6ce7ac3..92373d4 100644
--- a/couchdb/server/views.py
+++ b/couchdb/server/views.py
@@ -3,7 +3,7 @@
 import copy
 import logging
 
-from couchdb import json
+from couchdb import json, util
 from couchdb.server.exceptions import QueryServerException, Error
 
 __all__ = ('map_doc', 'reduce', 'rereduce')
@@ -37,7 +37,21 @@ def map_doc(server, doc):
         for idx, func in enumerate(server.state['functions']):
             # TODO: https://issues.apache.org/jira/browse/COUCHDB-729
             # Apply copy.deepcopy for `key` and `value` to fix this issue
-            _append([[key, value] for key, value in func(doc) or []])
+            pairs = []
+            for pair in func(doc) or []:
+                # avoid str types being unpack without error
+                if isinstance(pair, util.strbase):
+                    raise Error('map function must yield/return key-value pairs'
+                                ', not a string: `{0!r}`'.format(pair))
+                try:
+                    key, val = pair
+                except (TypeError, ValueError):
+                    raise Error('map function must yield/return key-value pairs'
+                                ', invalid value: `{0!r}`'.format(pair))
+                else:
+                    pairs.append([key, val])
+            _append(pairs)
+
             if doc != orig_doc:
                 log.warning('Document `%s` had been changed by map function'
                             ' `%s`, but was restored to original state',
diff --git a/couchdb/tests/server/views.py b/couchdb/tests/server/views.py
index eab81a0..0788d4c 100644
--- a/couchdb/tests/server/views.py
+++ b/couchdb/tests/server/views.py
@@ -117,6 +117,92 @@ class MapTestCase(unittest.TestCase):
         doc = {'_id': 'foo'}
         views.map_doc(self.server, doc)
 
+    def test_yield_non_iterable(self):
+        """should raise Error if map function do not yield iterable"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  yield 2, 4\n'
+            '  yield 42'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('invalid value' in err.args[0])
+            self.assertTrue('`42`' in err.args[0])
+
+    def test_yield_non_pair(self):
+        """should raise Error if map function do not yield pair"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  yield 2, 4\n'
+            '  yield [42]'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('invalid value' in err.args[0])
+            self.assertTrue('`[42]`' in err.args[0])
+
+    def test_yield_str_type(self):
+        """should raise Error if map function yield a string"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  yield 2, 4\n'
+            '  yield "42"'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('not a string' in err.args[0])
+
+    def test_return_non_iterable(self):
+        """should raise Error
+        if map function do not return a serise of iterable"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  return [(2, 4), 42]'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('invalid value' in err.args[0])
+            self.assertTrue('`42`' in err.args[0])
+
+    def test_return_non_pair(self):
+        """should raise Error if map function return a serise contain non-pair"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  return [(2, 4), [42]]'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('invalid value' in err.args[0])
+            self.assertTrue('[42]' in err.args[0])
+
+    def test_return_str_type(self):
+        """should raise Error if map function return a serise contain string"""
+        state.add_fun(
+            self.server,
+            'def mapfun(doc):\n'
+            '  return [(2, 4), "42"]'
+        )
+        doc = {'_id': 'foo'}
+        try:
+            views.map_doc(self.server, doc)
+        except exceptions.Error as err:
+            self.assertTrue('not a string' in err.args[0])
+
```